clients/src/main/java/org/apache/kafka/clients/admin/NewPartitions.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Describes new partitions for a particular topic in a call to {@link AdminClient#createPartitions(Map)}.
+ *
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+public class NewPartitions {
+
+    private int totalCount;
+
+    private List<List<Integer>> newAssignments;
+
+    private NewPartitions(int totalCount, List<List<Integer>> newAssignments) {
+        this.totalCount = totalCount;
+        this.newAssignments = newAssignments;
```

Comment:
We should validate the parameters and we should have tests for that.

Reply:
I tried putting validation here. Obviously we can't validate properly because we don't know the replication factor or existing number of partitions.

We still need to have the server-side code for validating these things (non-Java clients), so I reflectively created invalid NewPartitions in the `AdminClientIntegrationTest` in order to test that. But that reveals a new problem: The validation in `NewPartitions` applies during deserialization of the `CreatePartitionsRequest`, which prevents the request making it as far as `KafkaApis` and `AdminManager` where "proper" validation (with knowledge of the current number of partitions, replication factor etc) occurs. Instead, exceptions propagating during deserialization seem to lead to the broker disconnecting from the client, presumably because it considers an exception thrown during deserialization to be an unrecoverable client error.

I'm minded to say we should just let the server do the validation, but maybe you have a better idea, @ijuma?

Reply:
That's a good point about validation causing issues when the broker tries to deserialize. I was just thinking that we'd validate that `newAssignments` was consistent, i.e. every inner list has the same size. Maybe we should do it in the static factory methods which are only used by the `AdminClient` instead of the broker.

Reply:
Doing it in the factory methods doesn't solve the problem though, because if we allow the broker a way to create a `NewPartitions` that avoids the validation, we necessarily expose that same mechanism to the client, defeating the point of having the validation in `NewPartitions`.

Reply:
Client-side validation never guarantees anything (people can always bypass them), this check would be simply to fail-fast in case of programming mistakes. In any case, if you think the benefit is not worth the extra complexity, we can leave it out.